File: tests/safe_mode/check_safe_mode.py
# Copyright 2019 PingCAP, Inc.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# See the License for the specific language governing permissions and
# limitations under the License.

"""script for checking safeMode in DML is as expected

In test case we send `update` SQL before, in and after a specific sharding DDL
procedure, and this script checks that update binlog event added during sharding
DDL not synced phase has safeMode true, otherwise the safeMode is false.
"""

#!/usr/bin/env python
# coding: utf-8

import json
import urllib
from collections import defaultdict


class TraceType(object):
    BinlogEvent = 1
    JobEvent = 2


class OpType(object):
    UPDATE = 2
    DDL = 4


class ReplicationEvent(object):
    QUERY_EVENT = 2
    UPDATE_ROWS_EVENTv0 = 21
    UPDATE_ROWS_EVENTv1 = 24
    UPDATE_ROWS_EVENTv2 = 31

    UPDATE = [UPDATE_ROWS_EVENTv0, UPDATE_ROWS_EVENTv1, UPDATE_ROWS_EVENTv2]


def scan_events():
    url = 'http://127.0.0.1:8264/events/scan?limit=1000'
    f = urllib.urlopen(url)
    return json.loads(f.read())


def _trace_id_getter(d):
    return d['event']['base']['traceID'].split('.')[0]


def _binlog_pos_with_name(ev):
    current_pos = ev['event']['state']['currentPos']
    return '.'.join([current_pos['name'], str(current_pos['pos'])])


def _event_tso(ev):
    return ev['event']['base']['tso']


def _check_safe_mode(expected, safe_mode, ev, start_ev, end_ev):
    if expected != safe_mode:
        raise Exception(
            ("safe_mode should be {}, got {}\nev: {}\nstart_evn:{}\nend_ev:{}\n").format(
            expected, safe_mode, ev, start_ev, end_ev))


def main():
    update_events = defaultdict(list)
    ddl_binlogs = defaultdict(list)
    ddl_events = defaultdict(dict)
    latest_ddl = defaultdict(list)
    for events in scan_events():
        for ev in events:
            if ev['type'] == TraceType.BinlogEvent:
                if ev['event']['opType'] == OpType.UPDATE and ev['event']['eventType'] in ReplicationEvent.UPDATE:
                    update_events[_trace_id_getter(ev)].append(ev)
                if ev['event']['opType'] == OpType.DDL:
                    trace_id = _trace_id_getter(ev)
                    pos = _binlog_pos_with_name(ev)
                    ddl_binlogs[trace_id].append(pos)
                    ddl_events[trace_id][pos] = ev

    # sort and get last sharding DDL tracing events
    for trace_id, binlogs in ddl_binlogs.items():
        binlogs = sorted(set(binlogs))
        evs = ddl_events[trace_id]
        latest_ddl[trace_id].append(evs[binlogs[-2]])
        latest_ddl[trace_id].append(evs[binlogs[-1]])

    # check safe mode of each `update` binlog event is reasonable
    count = 0
    for worker, evs in latest_ddl.items():
        start = _binlog_pos_with_name(evs[0])
        end = _binlog_pos_with_name(evs[1])
        for ev in update_events[worker]:
            pos = _binlog_pos_with_name(ev)
            safe_mode = ev['event']['state'].get('safeMode', False)
            # only DML added during sharding DDL not synced has safeMode true
            if pos > end or (pos >= start and _event_tso(ev) > _event_tso(evs[1])):
                _check_safe_mode(False, safe_mode, ev, evs[0], evs[1])
            elif pos >= start:
                _check_safe_mode(True, safe_mode, ev, evs[0], evs[1])
            count += 1
    print 'check {} update events passed!'.format(count)


if __name__ == '__main__':
    main()

File: tests/_dmctl_tools/check_safe_mode.go
// Copyright 2019 PingCAP, Inc.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// See the License for the specific language governing permissions and
// limitations under the License.
//
// script used for checking safeMode in DML is as expected
//
// In test case we send `update` SQL before, in and after a specific sharding DDL
// procedure, and this script checks that update binlog event added during sharding
// DDL not synced phase has safeMode true, otherwise the safeMode is false.
package main

import (
	"encoding/json"
	"fmt"
	"io/ioutil"
	"net/http"
	"os"
	"sort"
	"strings"

	"github.com/gogo/protobuf/jsonpb"
	"github.com/pingcap/errors"
	"github.com/siddontang/go-mysql/mysql"

	"github.com/pingcap/dm/dm/pb"
	"github.com/pingcap/dm/dm/tracer"
)

const (
	opTypeUpdate = 2
	opTypeDDL    = 4
)

func scan_events(offset, limit int) ([]byte, error) {
	uri := fmt.Sprintf("http://127.0.0.1:8264/events/scan?offset=%d&limit=%d", offset, limit)
	resp, err := http.Get(uri)
	if err != nil {
		return nil, errors.Trace(err)
	}
	defer resp.Body.Close()
	body, err := ioutil.ReadAll(resp.Body)
	if err != nil {
		return nil, errors.Trace(err)
	}

	return body, nil
}

func traceSourceExtract(ev *pb.SyncerBinlogEvent) string {
	return strings.Split(ev.Base.TraceID, ".")[0]
}

func binlogPosition(ev *pb.SyncerBinlogEvent) mysql.Position {
	return mysql.Position{
		Name: ev.State.CurrentPos.Name,
		Pos:  ev.State.CurrentPos.Pos,
	}
}

func checkSafeMode(obtained, expected bool, ev, startEv, endEv *pb.SyncerBinlogEvent) {
	if obtained != expected {
		panic(fmt.Sprintf("safe_mode should be %v, got %v\nev: %v\nstart_ev: %v\nend_ev: %v\n",
			expected, obtained, ev, startEv, endEv))
	}
}

type BinlogEventSlice []*pb.SyncerBinlogEvent

func (c BinlogEventSlice) Len() int {
	return len(c)
}

func (c BinlogEventSlice) Swap(i, j int) {
	c[i], c[j] = c[j], c[i]
}

func (c BinlogEventSlice) Less(i, j int) bool {
	posi := binlogPosition(c[i])
	posj := binlogPosition(c[j])
	return posi.Compare(posj) < 0
}

func main() {
	var (
		raw    []byte
		err    error
		offset int
		limit  int = 10
		count  int
	)
	checkInstance := fmt.Sprintf("mysql-replica-0%s", os.Args[1])
	updateEvents := make(map[string][]*pb.SyncerBinlogEvent)
	ddlEvents := make(map[string][]*pb.SyncerBinlogEvent)
	latestDDL := make(map[string][]*pb.SyncerBinlogEvent)
	fmt.Printf("check safe mode for instance %s\n", checkInstance)
	for {
		raw, err = scan_events(offset, limit)
		if err != nil {
			panic(errors.ErrorStack(err))
		}
		var events [][]tracer.TraceEvent
		err = json.Unmarshal(raw, &events)
		if err != nil {
			panic(err)
		}

		for _, event_group := range events {
			for _, e := range event_group {
				if e.Type == pb.TraceType_BinlogEvent {
					ev := &pb.SyncerBinlogEvent{}
					jsonE, _ := json.Marshal(e.Event)
					err = jsonpb.UnmarshalString(string(jsonE), ev)
					if err != nil {
						panic(err)
					}
					worker := traceSourceExtract(ev)
					if ev.OpType == opTypeUpdate {
						if _, ok2 := updateEvents[worker]; !ok2 {
							updateEvents[worker] = make([]*pb.SyncerBinlogEvent, 0)
						}
						updateEvents[worker] = append(updateEvents[worker], ev)
					} else if ev.OpType == opTypeDDL {
						if _, ok2 := ddlEvents[worker]; !ok2 {
							ddlEvents[worker] = make([]*pb.SyncerBinlogEvent, 0)
						}
						ddlEvents[worker] = append(ddlEvents[worker], ev)
					}
				}
			}
		}

		offset += len(events)
		if len(events) < limit {
			break
		}
	}

	// sort and get last sharding DDL tracing events
	for worker, events := range ddlEvents {
		sort.Sort(BinlogEventSlice(events))
		latestDDL[worker] = make([]*pb.SyncerBinlogEvent, 2)
		// we check the second round of sharding DDL only
		if len(events) < 2 {
			panic(fmt.Sprintf("invalid ddl events count, events: %v", events))
		}
		latestDDL[worker][1] = events[len(events)-1]
		endPos := binlogPosition(latestDDL[worker][1])
		for i := len(events) - 2; i >= 0; i-- {
			pos := binlogPosition(events[i])
			if pos.Compare(endPos) != 0 {
				latestDDL[worker][0] = events[i]
				break
			}
		}
		if latestDDL[worker][0] == nil {
			panic(fmt.Sprintf("first sharding DDL not found, events: %v", events))
		}
	}

	// check safe mode of each `update` binlog event is reasonable
	for worker, evs := range latestDDL {
		if worker != checkInstance {
			continue
		}
		startPos := binlogPosition(evs[0])
		endPos := binlogPosition(evs[1])
		for _, ev := range updateEvents[worker] {
			pos := binlogPosition(ev)
			safeMode := ev.State.SafeMode
			if pos.Compare(endPos) > 0 || (pos.Compare(startPos) >= 0 && ev.Base.Tso > evs[1].Base.Tso) {
				checkSafeMode(safeMode, false, ev, evs[0], evs[1])
			} else if pos.Compare(startPos) >= 0 {
				checkSafeMode(safeMode, true, ev, evs[0], evs[1])
			}
			count += 1
		}
	}

	fmt.Printf("check %d update events passed!\n", count)
}
